File: nodejs/src/info.ts
import type { Network, Token } from "./types.js"

interface RawToken {
  address: string
  symbol: string
  decimals: number
  active: boolean
}

interface InfoResponse {
  status: "success" | "error"
  data?: Record<string, RawToken[] | undefined>
}

export async function fetchTokens(network: Network, infoUrl: string): Promise<Token[]> {
  return fetchTokensFrom(network, infoUrl)
}

export async function fetchTokensFrom(network: Network, url: string): Promise<Token[]> {
  const fullUrl = `${url}?network=${network}`

  let res: Response
  try {
    res = await fetch(fullUrl)
  } catch (e) {
    throw new Error(`AFI info request failed: ${(e as Error).message}`)
  }

  if (!res.ok) throw new Error(`AFI info HTTP ${res.status}`)

  const json = (await res.json()) as InfoResponse

  if (json.status !== "success" || !json.data) {
    throw new Error(`AFI info returned no data for network: ${network}`)
  }

  const raw = json.data[network]
  if (!raw || raw.length === 0) {
    throw new Error(`AFI info returned no tokens for network: ${network}`)
  }

  return raw.map((t) => ({
    address:  t.address as `0x${string}`,
    symbol:   t.symbol,
    decimals: t.decimals,
    active:   t.active,
  }))
}

File: go/info.go
package afi

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/ethereum/go-ethereum/common"
)

type infoResponse struct {
	Status string                `json:"status"`
	Data   map[string][]rawToken `json:"data"`
}

type rawToken struct {
	Address  string `json:"address"`
	Symbol   string `json:"symbol"`
	Decimals uint8  `json:"decimals"`
	Active   bool   `json:"active"`
}

// fetchTokens fetches available tokens from the AFI info endpoint for the given network.
func fetchTokens(ctx context.Context, network Network, infoURL string) ([]Token, error) {
	return fetchTokensFrom(ctx, network, infoURL)
}

// fetchTokensFrom allows injecting a custom URL — used by tests.
func fetchTokensFrom(ctx context.Context, network Network, url string) ([]Token, error) {
	u := url + "?network=" + string(network)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, fmt.Errorf("build info request: %w", err)
	}

	httpClient := &http.Client{Timeout: 15 * time.Second}
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("info request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("info HTTP %d", resp.StatusCode)
	}

	var result infoResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decode info: %w", err)
	}

	if result.Status != "success" {
		return nil, fmt.Errorf("no %s tokens in info response", network)
	}

	networkTokens, ok := result.Data[string(network)]
	if !ok || len(networkTokens) == 0 {
		return nil, fmt.Errorf("no %s tokens in info response", network)
	}

	tokens := make([]Token, len(networkTokens))
	for i, t := range networkTokens {
		tokens[i] = Token{
			Address:  common.HexToAddress(t.Address),
			Symbol:   t.Symbol,
			Decimals: t.Decimals,
			Active:   t.Active,
		}
	}
	return tokens, nil
}
